services/brig/src/Brig/Provider/API.hs

```diff
@@ -790,69 +856,93 @@ removeBotH (zusr ::: zcon ::: cid ::: bid) = do
     p <- lift $ User.lookupUser (botUserId bot)
-    maybe (throwStd userNotFound) (return . json . publicProfile) p
+    maybe (throwStd userNotFound) (return . publicProfile) p
 
 botGetClientH :: BotId -> Handler Response
 botGetClientH bot = do
-    c <- lift $ listToMaybe <$> User.lookupClients (botUserId bot)
-    maybe (throwStd clientNotFound) (return . json) c
+    maybe (throwStd clientNotFound) (pure . json) =<< lift (botGetClient bot)
+
+botGetClient :: BotId -> AppIO (Maybe Client)
+botGetClient bot = do
+    listToMaybe <$> User.lookupClients (botUserId bot)
 
 botListPrekeysH :: BotId -> Handler Response
 botListPrekeysH bot = do
+    json . fromMaybe [] <$> botListPrekeys bot
+
+botListPrekeys :: BotId -> Handler (Maybe [PrekeyId])
+botListPrekeys bot = do
     clt <- lift $ listToMaybe <$> User.lookupClients (botUserId bot)
     case clientId <$> clt of
-        Nothing -> return $ json ([] :: [PrekeyId])
-        Just ci -> json <$> lift (User.lookupPrekeyIds (botUserId bot) ci)
+        Nothing -> return Nothing
+        Just ci -> Just <$> lift (User.lookupPrekeyIds (botUserId bot) ci)
 
 botUpdatePrekeysH :: BotId ::: JsonRequest UpdateBotPrekeys -> Handler Response
 botUpdatePrekeysH (bot ::: req) = do
-    upd <- parseJsonBody req
+    empty <$ (botUpdatePrekeys bot =<< parseJsonBody req)
+
+botUpdatePrekeys :: BotId -> UpdateBotPrekeys -> Handler ()
+botUpdatePrekeys bot upd = do
     clt <- lift $ listToMaybe <$> User.lookupClients (botUserId bot)
     case clt of
         Nothing -> throwStd clientNotFound
         Just  c -> do
             let pks = updateBotPrekeyList upd
             User.updatePrekeys (botUserId bot) (clientId c) pks !>> clientDataError
-    return empty
 
 botClaimUsersPrekeysH :: JsonRequest UserClients -> Handler Response
 botClaimUsersPrekeysH req = do
-    body <- parseJsonBody req
+    json <$> (botClaimUsersPrekeys =<< parseJsonBody req)
+
+botClaimUsersPrekeys :: UserClients -> Handler (UserClientMap (Maybe Prekey))
+botClaimUsersPrekeys body = do
     maxSize <- fromIntegral . setMaxConvSize <$> view settings
     when (Map.size (userClients body) > maxSize) $
         throwStd tooManyClients
-    json <$> lift (Client.claimMultiPrekeyBundles body)
+    lift (Client.claimMultiPrekeyBundles body)
 
 botListUserProfilesH :: List UserId -> Handler Response
 botListUserProfilesH uids = do
+    json <$> botListUserProfiles uids
+
+botListUserProfiles :: List UserId -> Handler [Ext.BotUserView]
+botListUserProfiles uids = do
     us <- lift $ User.lookupUsers (fromList uids)
-    return (json (map mkBotUserView us))
+    return (map mkBotUserView us)
 
 botGetUserClientsH :: UserId -> Handler Response
-botGetUserClientsH uid =
-    json <$> lift (fmap pubClient <$> User.lookupClients uid)
+botGetUserClientsH uid = do
+    json <$> botGetUserClients uid
+
+botGetUserClients :: UserId -> Handler [PubClient]
+botGetUserClients uid = do
+    lift (pubClient <$$> User.lookupClients uid)
```

Comment:
The `lift` could be pulled out of this function.

Reply:
117118cb69c5791346aa01a36a7547a9424016e2